✨Create new extension - AMP-smartlinks

PhilWinchester opened the pull request:
Resolves #19470

---

## History
This PR was initially merged here (https://github.com/ampproject/amphtml/pull/20494) and then reverted here (https://github.com/ampproject/amphtml/pull/20891) because it broke master due to it's reliance on another extension. The changes in this PR are the same as `20494`, with the addition to `compile.js` described here: https://github.com/ampproject/amphtml/pull/20494#issuecomment-465671826

---

## What it does
Create an extension that lets Narrativ offer our publishers our Linkmate service in AMP pages. The extension scans the page for any outbound links the publisher wants targeted and sends them to our API for monetization. Upon API response we use the `link-rewriter` service to map that response to the links in the initial payload. When a user clicks one of the monetized links we swap the href with the monetized link. More about Linkmate [here](http://docs.narrativ.com/en/stable/linkmate.html#linkmate).

Gregable commented on extensions/amp-smartlinks/validator-amp-smartlinks.protoascii:
```diff
@@ -0,0 +1,56 @@
+    name: "linkmate"
+    mandatory: true
+    value: "false"
+    value: "true"
```
Rather than `linkmate=true` / `linkmate=false`, have you considered instead using the presence of the attribute as the true/false indication. This would look like:

```
attrs: {
  name: "linkmate"
  value: ""
}
```

This requires a null value, but doesn't mandate the attribute presence. This is a more commonly seen pattern in HTML. Think `<button>` / `<button disabled>` rather than `<button disabled=false>` / `<button disabled=true>`.

c-nichols replied:
I can see that -- I think this will likely change in the future, as we move to consolidate functionality with a couple similar extensions.  Would you be open to merging as-is?  Most of our customers are going to take our auto-generated code anyways, and we'd really like to get this merged.

honeybadgerdontcare replied:
We prefer to avoid making something valid for only a short time then having to go through a lengthy deprecation process to make it invalid. It would be better if it can be valid from the start.

zhouyx replied:
@c-nichols even if we are going to move to the more generic approach, we will need to keep the existing amp-smartlinks extension for backward compatibility. Do you see any plan to support linkmate value other than `true/false` in the future.  

@Gregable and @honeybadgerdontcare I personally think the current validator spec is the stricter version. We can add `value: ""` on top of `value: "true"` and `value: "false"` without breaking existing page. WDYT

Gregable commented on extensions/amp-smartlinks/amp-smartlinks.md:
```diff
@@ -0,0 +1,102 @@
+  <tr>
+    <td class="col-fourty"><strong>exclusive-links</strong></td>
+    <td><strong>Optional</strong></td>
+    <td>Field to flag if you want exclusive links on an article. Expected values are <code>true</code> or <code>false</code>.</td>
```
The protoascii encodes that the only allowed value is an empty string.  I think that's the right approach (see my comment in the protoascii file), but this document doesn't match up.

Gregable commented on extensions/amp-smartlinks/amp-smartlinks.md:
```diff
@@ -0,0 +1,102 @@
+  <tr>
+    <td class="col-fourty"><strong>link-attribute</strong></td>
+    <td><strong>Optional</strong></td>
+    <td>If you store the "plain" url for a link in a different element attribute than <code>href</code> you can specify so here. Default value: <code>href</code>.</td>
```
I don't think this is really possible with AMP, since custom javascript can't be used. Does it make sense to remove this to streamline the API for your user?

zhouyx replied:
I think replacing different attribute other than `href` for anchor tag is possible, however I don't find replacing attribute other than `href` to be useful.  @PhilWinchester what is an example value for `link-attribute`?

c-nichols replied:
@zhouyx @Gregable sometimes we build on top of other providers' solutions.  In this case, the `href` attribute may have been modified server-side to include a 3rd-party link wrapper, and the original url would be found in an attribute like `data-original-url`.

Full example might be something like:
```
<a href="https://thirdparty.com/redirect?url=destination.com" data-original-url="https://destination.com">
```

In this case, we would want to pull the original url from the data-original-url attribute, to run through our own transformations, and finally insert into `href`.

Gregable replied:
Understood, seems reasonable.

PhilWinchester commented:
Hi @zhouyx @alabiaga,

I re-opened my changes here if you don't mind taking another look 😃. Added your comment [here](https://github.com/ampproject/amphtml/pull/20967/files#diff-b2216d09c55c9ee1f2f2c879bee3b619R245).

PhilWinchester commented:
@zhouyx @alabiaga 

Sorry to pester you guys all over again, just hoping to close out this PR as quickly as reasonable!

Thanks, Phil

zhouyx commented:
Thank you @PhilWinchester Have tested locally with compiled file
```
gulp dist --extensions=amp-smartlinks
gulp serve --compiled
```

PhilWinchester commented:
@zhouyx Yep! I tested with those commands and `gulp dist --fortesting`. No errors came up.

c-nichols commented:
@Gregable @zhouyx given this was previously merged, what would you like addressed before re-merging vs. in a followup?

Gregable commented:
@c-nichols I didn't notice that this was previously merged. Apologies for the moving goalposts.

As @honeybadgerdontcare mentioned, it might be difficult to make these changes later if there is observed usage by publishers, as we wouldn't want to make a breaking change.

I'd lean towards making the one change I recommended as my guess is that the modification is small. However, If you feel comfortable with the risk that you will be stuck with the current syntax, I'm OK with that. Consider my comment advisory. I can't speak for @zhouyx

c-nichols commented:
Thanks @gregable - we will make that change and then hopefully final approvals will come in quickly ;)

PhilWinchester commented:
Hi @alabiaga & @zhouyx ,

I've updated the validator per conversation above and addressed the issue from the previous PR. Is it possible to get a quick review on this and close these changes out?

alabiaga commented on extensions/amp-smartlinks/0.1/amp-smartlinks.js:
```diff
@@ -0,0 +1,218 @@
+    const options = {linkSelector: this.linkmateOptions_.linkSelector};
+
+    return this.linkRewriterService_.registerLinkRewriter(
+        SMARTLINKS_REWRITER_ID,
```
nit remove from constants.js and export a const named TAG in this class as we do in other extensions and use this in all places where 'amp-smartlinks' string is used. 

    export TAG = 'amp-smartlinks';

You can address in a separate PR.

PhilWinchester replied:
Makes sense. I added it here as it's a small change.

alabiaga commented on extensions/amp-smartlinks/0.1/linkmate-options.js:
```diff
@@ -0,0 +1,93 @@
+ */
+function getNrtvAccountName_(element) {
+  const nrtvSlug = element.getAttribute('nrtv-account-name');
+  userAssert(nrtvSlug,
```
there is a validator rule that would catch this so the userAssert shouldn't be required.

alabiaga replied:
addressing in a separate PR?

PhilWinchester replied:
Whoops completely misread this comment initially. I'll update the PR with to remove this assert.

PhilWinchester replied:
updated

zhouyx commented on extensions/amp-smartlinks/0.1/linkmate-options.js:
```diff
@@ -52,11 +52,9 @@ function getNrtvAccountName_(element) {
+  const linkmateEnabled = element.getAttribute('linkmate');
 
-  return linkmateEnabled === 'true';
+  return !!linkmateEnabled && linkmateEnabled.toLowerCase() === 'true';
```
Given the new validation rule, I believe `linkmateEnabled` can never be a string. 
```
attrs: {
    name: "linkmate"
    value: ""
  }	
```
@honeybadgerdontcare @Gregable to confirm

If so the check will look like `return !!element.hasAttribute('linkmate')`. And you'll need to change your example and test files as well.

honeybadgerdontcare replied:
It can be a string. It can be `"linkmateEnabled"` or empty string.

zhouyx replied:
@honeybadgerdontcare can it be arbitrary string? What about
```
<amp-smartlinks linkmate> //empty string
```
```
<amp-smartlinks linkmate="linkmateEnabled"> //arbitrary string
```
```
<amp-smartlinks linkmate="true"> // "true"
```
```
<amp-smartlinks linkmate=true> // boolean
```

honeybadgerdontcare replied:
`value: ""` means the attribute is only valid when it does not have a value or when that value is the same as the attribute name:

```
<amp-smartlinks linkmate></amp-smartlinks> // attribute valid
<amp-smartlinks linkmate=linkmate></amp-smartlinks> // attribute valid
<amp-smartlinks linkmate="true"></amp-smartlinks> // attribute invalid
<amp-smartlinks linkmate=true></amp-smartlinks> // attribute invalid
```

Side note that `=true` and `="true"` are the same thing to the validator, the value is the string `true`.

zhouyx replied:
Thank you @honeybadgerdontcare for the clarification! @PhilWinchester looks like `hasAttribute('linkmate')` is the right way to go.

PhilWinchester commented:
Thanks for the quick feedback @zhouyx @alabiaga. I've addressed your comments if you don't mind taking another look!

zhouyx commented on extensions/amp-smartlinks/0.1/test/test-amp-smartlinks.js:
```diff
@@ -0,0 +1,247 @@
+
+          const smartlinkOptions = {
+            'nrtv-account-name': 'thisisnotapublisher',
+            'linkmate': true,
```
@PhilWinchester Since we change the validation rule, the test files will need to be updated as well.

PhilWinchester replied:
@zhouyx I wasn't sure the best way to mock an empty attribute tag in this dict so I went with empty strings. Let me know if that's an issue for this test.

zhouyx replied:
empty string should be fine.

zhouyx commented on extensions/amp-smartlinks/0.1/test/test-amp-smartlinks.js:
```diff
@@ -0,0 +1,246 @@
+
+          const smartlinkOptions = {
+            'nrtv-account-name': 'alwaysastring',
+            'linkmate': 1234,
```
nit: same

PhilWinchester replied:
We want this one here b/c it's a test asserting how we handle bad input. In this case we expect `linkmate` to still be true b/c we just check for it's presence in the tag.

zhouyx replied:
I see. That makes sense. I was curious why it was true before the validation change.

zhouyx commented on extensions/amp-smartlinks/0.1/test/test-amp-smartlinks.js:
```diff
@@ -0,0 +1,246 @@
+        it('Should fetch Linkmate Options from API', () => {
+          const options = {
+            'nrtv-account-name': 'testingconfigpub',
+            'linkmate': true,
```
ditto

zhouyx commented on extensions/amp-smartlinks/0.1/test/test-amp-smartlinks.js:
```diff
@@ -0,0 +1,246 @@
+        beforeEach(() => {
+          const options = {
+            'nrtv-account-name': 'thisisnotapublisher',
+            'linkmate': true,
```
ditto

zhouyx commented on extensions/amp-smartlinks/0.1/test/test-amp-smartlinks.js:
```diff
@@ -0,0 +1,246 @@
+        beforeEach(() => {
+          const options = {
+            'nrtv-account-name': 'thisisnotapublisher',
+            'linkmate': true,
```
ditto

zhouyx commented on extensions/amp-smartlinks/0.1/test/test-amp-smartlinks.js:
```diff
@@ -0,0 +1,246 @@
+        beforeEach(() => {
+          const options = {
+            'nrtv-account-name': 'thisisnotapublisher',
+            'linkmate': true,
```
ditto

PhilWinchester commented:
@alabiaga @zhouyx addressed your comments. If you could take another look 😄

alabiaga commented on extensions/amp-smartlinks/0.1/linkmate-options.js:
```diff
@@ -0,0 +1,86 @@
+ * @param {!Element} element
+ * @return {boolean}
+ */
+function getExclusiveLinksFlag_(element) {
```
should be hasExclusiveLinksFlag_ as you're checking if the attribute exists and not returning the attribute value.

zhouyx commented on extensions/amp-smartlinks/amp-smartlinks.md:
```diff
@@ -0,0 +1,102 @@
+  </tr>
+  <tr>
+    <td class="col-fourty"><strong>linkmate</strong></td>
+    <td><strong>Required</strong></td>
```
nit: the `linkmate` attribute is not longer required.

PhilWinchester commented:
@alabiaga @zhouyx Updated!

PhilWinchester commented:
Hi @alabiaga @zhouyx,

Is this bundle size failure something I should worry about? I haven't noticed this test failing until this latest push.

alabiaga commented:
@PhilWinchester let's pull from master and do another push for this PR to retrigger that failure.

c-nichols commented:
@alabiaga tests passed :fireworks: is this good to merge?
